boa_engine/src/builtins/promise/mod.rs

```diff
@@ -316,6 +317,247 @@ impl Promise {
+        let promise_capability = PromiseCapability::new(c, context)?;
+
+        // Note: We already checked that `C` is a constructor in `NewPromiseCapability(C)`.
+        let c = c.as_object().expect("must be a constructor");
+
+        // 3. Let promiseResolve be Completion(GetPromiseResolve(C)).
+        let promise_resolve = Self::get_promise_resolve(c, context);
+
+        // 4. IfAbruptRejectPromise(promiseResolve, promiseCapability).
+        if_abrupt_reject_promise!(promise_resolve, promise_capability, context);
+
+        // 5. Let iteratorRecord be Completion(GetIterator(iterable)).
+        let iterator_record = args.get_or_undefined(0).get_iterator(context, None, None);
+
+        // 6. IfAbruptRejectPromise(iteratorRecord, promiseCapability).
+        if_abrupt_reject_promise!(iterator_record, promise_capability, context);
+        let mut iterator_record = iterator_record;
+
+        // 7. Let result be Completion(PerformPromiseAll(iteratorRecord, C, promiseCapability, promiseResolve)).
+        let mut result = Self::perform_promise_all(
+            &mut iterator_record,
+            c,
+            &promise_capability,
+            &promise_resolve,
+            context,
+        )
+        .map(JsValue::from);
+
+        // 8. If result is an abrupt completion, then
+        if result.is_err() {
+            // a. If iteratorRecord.[[Done]] is false, set result to Completion(IteratorClose(iteratorRecord, result)).
+            if !iterator_record.done() {
+                result = iterator_record.close(result, context);
+            }
+
+            // b. IfAbruptRejectPromise(result, promiseCapability).
+            if_abrupt_reject_promise!(result, promise_capability, context);
+
+            return Ok(result);
+        }
+
+        // 9. Return ? result.
+        result
+    }
+
+    /// `PerformPromiseAll ( iteratorRecord, constructor, resultCapability, promiseResolve )`
+    ///
+    /// More information:
+    ///  - [ECMAScript reference][spec]
+    ///
+    /// [spec]: https://tc39.es/ecma262/#sec-performpromiseall
+    fn perform_promise_all(
+        iterator_record: &mut IteratorRecord,
+        constructor: &JsObject,
+        result_capability: &PromiseCapability,
+        promise_resolve: &JsObject,
+        context: &mut Context,
+    ) -> JsResult<JsObject> {
+        #[derive(Debug, Trace, Finalize)]
+        struct ResolveElementCaptures {
+            #[unsafe_ignore_trace]
+            already_called: Rc<Cell<bool>>,
+            index: usize,
+            values: GcCell<Vec<JsValue>>,
+            capability_resolve: JsFunction,
+            #[unsafe_ignore_trace]
+            remaining_elements_count: Rc<Cell<i32>>,
+        }
+
+        // 1. Let values be a new empty List.
+        let values = GcCell::new(Vec::new());
+
+        // 2. Let remainingElementsCount be the Record { [[Value]]: 1 }.
+        let remaining_elements_count = Rc::new(Cell::new(1));
```

Comment:
As far as I can tell, the reason behind this being inside cells is that we need to pass them as captures below, right? But we don't use them until step `j`, if I'm not mistaken. Wouldn't it be more efficient to wrap them in cells after the `loop` has completed?